compiler/src/iree/compiler/Dialect/Flow/Transforms/DetachElementwiseFromNamedOps.cpp

```diff
@@ -0,0 +1,125 @@
+//
+// Detaches elementwise ops from Linalg named ops in preparation for following
+// fusion and bufferization.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/StructuredOpsUtils.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+struct DetachElementwisePattern
+    : public OpInterfaceRewritePattern<linalg::ContractionOpInterface> {
+  using OpInterfaceRewritePattern::OpInterfaceRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::ContractionOpInterface interfaceOp,
+                                PatternRewriter &rewriter) const override {
+    auto linalgOp = cast<linalg::LinalgOp>(interfaceOp.getOperation());
+    if (!linalgOp.hasTensorSemantics()) return failure();
+
+    // Nothing to do if the output tensor operand is already a fill op.
+    linalg::OpOperandVector outputOperands = linalgOp.getOutputTensorOperands();
+    if (outputOperands.size() != 1) return failure();
+    Value outputOperand = outputOperands.front()->get();
+    if (outputOperand.getDefiningOp<linalg::FillOp>()) return failure();
+
+    auto outputType = outputOperand.getType().cast<RankedTensorType>();
+    if (!outputType.getElementType().isIntOrFloat()) return failure();
+    auto elementType = outputType.getElementType();
+
+    Location loc = interfaceOp.getLoc();
+
+    // Create a zero tensor as the new output tensor operand to the Linalg
+    // contraction op.
+    SmallVector<Value> dynamicDims;
+    for (unsigned i = 0; i < outputType.getRank(); i++) {
+      if (outputType.isDynamicDim(i))
+        dynamicDims.push_back(
+            rewriter.create<tensor::DimOp>(loc, outputOperand, i));
+    }
+    auto initOp = rewriter.create<linalg::InitTensorOp>(
+        loc, dynamicDims, outputType.getShape(), elementType);
+    Value zero = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(elementType));
+    Value fill =
+        rewriter.create<linalg::FillOp>(loc, zero, initOp.result()).result();
+
+    // Update the contraction op to use the new zero tensor as output operand.
+    rewriter.updateRootInPlace(linalgOp,
+                               [&]() { linalgOp.setOutputOperand(0, fill); });
+
+    // Create a generic op to add back the original output tensor operand.
+    rewriter.setInsertionPointAfter(linalgOp);
+    auto identityMap = AffineMap::getMultiDimIdentityMap(outputType.getRank(),
```

Comment:
Might be safer to use the indexing map for the `outs` operand from the root op.

Reply:
Your right about that part, but this is making an assumption that output is always computed using identity map of the parallel dimension. It could be a permutation of the identity map as well.
If you want to keep it simple you can just fail on the indexing map of the `outs` operand not being identity w.r.t to parallel dimensions. WIthout that this will silently miscompile

Reply:
Yeah that's a good point; missed that. Tightened it a bit.